onnxruntime/core/providers/nuphar/mti_x86/tensor/scatter.cc

```diff
@@ -0,0 +1,210 @@
+    }
+  }
+
+  // copy input data into output
+  int64_t input_size = DLTensorSize(input);
+  memcpy(static_cast<char*>(output->data) + output->byte_offset,
+         static_cast<char*>(input->data) + input->byte_offset,
+         input_size * input->dtype.bits / 8);
+
+  // Record the size of each input dimension, i.e. the number of
+  // elements contained by a single element of current dimension.
+  // For example, for input[3][4][5][6], dimension_sizes will be
+  // [4*5*6, 5*6, 6, 1], i.e. [120, 30, 6, 1]
+  std::vector<int64_t> dimension_sizes(num_dims, 0);
+  dimension_sizes[num_dims - 1] = 1;
+  for (int64_t i = num_dims - 2; i >= 0; i--) {
+    dimension_sizes[i] = dimension_sizes[i+1] * input->shape[i+1];
+  }
+
+  T* output_data = reinterpret_cast<T*>(static_cast<char*>(output->data) + output->byte_offset);
+  T* updates_data = reinterpret_cast<T*>(static_cast<char*>(updates->data) + updates->byte_offset);
+  const std::vector<int> indices_shape(indices->shape, indices->shape + num_dims);
+  // Because indices data is flat, running_indices maintains indices's original dimensions.
+  // We will use its dimensions to compute the corresponding index (or offset) to output_data,
+  // which is also flat.
+  std::vector<int64_t> running_indices(num_dims, 0);
+
+  for (int64_t i = 0; i < indices_size; i++) {
+    Tind idx = indices_data_vec[i];
+    // output indices come from running_indices
+    std::vector<int64_t> curr_output_indices = running_indices;
+    curr_output_indices[axis] = static_cast<int64_t>(idx);
+
+    // get the index into output_data
+    int64_t output_idx = 0;
+    for (int j = 0; j < num_dims; j++) {
+      output_idx += curr_output_indices[j] * dimension_sizes[j];
+    }
+
+    // update data
+    output_data[output_idx] = updates_data[i];
+
+    // update running_indices
+    Tind carry = 1;
+    for (int j = num_dims - 1; j >= 0; j--) {
+      if (carry == 0) break;
+      Tind curr_idx = running_indices[j] + carry;
+      running_indices[j] = curr_idx % indices_shape[j];
+      carry = curr_idx / indices_shape[j];
+    }
+  }
+}
+
+#define REGISTER_EXTERN_SCATTER(input_type, index_type)               \
+  TVM_REGISTER_GLOBAL(GET_EXTERN_SCATTER_STR(input_type, index_type)) \
+      .set_body([](tvm::TVMArgs args, tvm::TVMRetValue* ret) {        \
+        ScatterCommon<input_type, index_type>(args, ret);             \
+      });
+
+#define REGISTER_EXTERN_SCATTER_PAIR(input_type) \
+  REGISTER_EXTERN_SCATTER(input_type, int32_t)   \
+  REGISTER_EXTERN_SCATTER(input_type, int64_t)   \
+
+REGISTER_EXTERN_SCATTER_PAIR(bool)
+REGISTER_EXTERN_SCATTER_PAIR(int8_t)
+REGISTER_EXTERN_SCATTER_PAIR(uint8_t)
+REGISTER_EXTERN_SCATTER_PAIR(int16_t)
+REGISTER_EXTERN_SCATTER_PAIR(uint16_t)
+REGISTER_EXTERN_SCATTER_PAIR(int32_t)
+REGISTER_EXTERN_SCATTER_PAIR(uint32_t)
+REGISTER_EXTERN_SCATTER_PAIR(int64_t)
+REGISTER_EXTERN_SCATTER_PAIR(uint64_t)
+REGISTER_EXTERN_SCATTER_PAIR(float)
+REGISTER_EXTERN_SCATTER_PAIR(double)
```

Comment:
I think we may reduce the templates based on element size of int8/int16/int32/int64, since scatter does not involve computation and just do copy.

Reply:
I thought about that too, but I felt that the current way is more readable, because there is one-to-one map between a parameter and a valid C++ type.